printer/printer_test.go

```diff
@@ -0,0 +1,179 @@
+package printer
+
+import (
+	"testing"
+
+	qt "github.com/frankban/quicktest"
+	"github.com/warpfork/go-wish"
+
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/fluent/qp"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+func TestSimpleData(t *testing.T) {
+	n, _ := qp.BuildMap(basicnode.Prototype.Any, -1, func(ma datamodel.MapAssembler) {
+		qp.MapEntry(ma, "some key", qp.String("some value"))
+		qp.MapEntry(ma, "another key", qp.String("another value"))
+		qp.MapEntry(ma, "nested map", qp.Map(2, func(ma datamodel.MapAssembler) {
+			qp.MapEntry(ma, "deeper entries", qp.String("deeper values"))
+			qp.MapEntry(ma, "more deeper entries", qp.String("more deeper values"))
+		}))
+		qp.MapEntry(ma, "nested list", qp.List(2, func(la datamodel.ListAssembler) {
+			qp.ListEntry(la, qp.Int(1))
+			qp.ListEntry(la, qp.Int(2))
+		}))
+	})
+	qt.Check(t, Sprint(n), qt.CmpEquals(), wish.Dedent(`
+		map{
+			string{"some key"}: string{"some value"}
+			string{"another key"}: string{"another value"}
+			string{"nested map"}: map{
+				string{"deeper entries"}: string{"deeper values"}
+				string{"more deeper entries"}: string{"more deeper values"}
+			}
+			string{"nested list"}: list{
+				0: int{1}
+				1: int{2}
+			}
+		}`,
+	))
+}
+
+func TestTypedData(t *testing.T) {
+	t.Run("structs", func(t *testing.T) {
+		type FooBar struct {
+			Foo string
+			Bar string
+		}
+		ts := schema.MustTypeSystem(
+			schema.SpawnString("String"),
+			schema.SpawnStruct("FooBar", []schema.StructField{
+				schema.SpawnStructField("foo", "String", false, false),
+				schema.SpawnStructField("bar", "String", false, false),
+			}, nil),
+		)
+		n := bindnode.Wrap(&FooBar{"x", "y"}, ts.TypeByName("FooBar"))
+		qt.Check(t, Sprint(n), qt.CmpEquals(), wish.Dedent(`
+			struct<FooBar>{
+				foo: string<String>{"x"}
```

Comment:
`string<String>` is getting weird, I don't suppose it's reasonable to special-case the typed data model kinds like `String` to omit the typing here?

Reply:
Yeah I'm on the fence on this.  At first I thought I'd strip it, as you suggest.  Then I looked at the printout the first time I got one to come out, and was like... hum: typed vs untyped string seems like the kind of distinction that could actually trip someone up someday, so maybe that's useful.

I'm _still_ on the fence, a couple days later.  I can't think of _how_ typed-vs-untyped string would trip someone up, exactly, either.  It's just a nagging fear.

Maybe a config option for "`SmartElidePreludeTypeInfo bool`"?

Reply:
I think I'd propose we leave it in verbose mode for now, someday add that config flag, and then if enough people are perturbed by the verbosity we flip the default of that flag to true.  We should probably get some impressions from people further out from us before deciding to elide things that might be obvious to us but nonobvious to new users, is my thinking.

Reply:
Added a config placeholder for this and attached documentation for it so someone looking for such a feature can at least find that we've considered it.  Haven't implemented it immediately, however.